esp-mbedtls/src/lib.rs

```diff
@@ -87,10 +88,83 @@ pub fn set_debug(level: u32) {
     }
 }
 
+/// Holds a X509 certificate
+///
+/// # Examples
+/// Initialize with a PEM certificate
+/// ```
+/// const CERTIFICATE: &[u8] = include_bytes!("certificate.pem");
+/// let cert = X509::pem(CERTIFICATE).unwrap();
+/// ```
+///
+/// Initialize with a DER certificate
+/// ```
+/// const CERTIFICATE: &[u8] = include_bytes!("certificate.der");
+/// let cert = X509::der(CERTIFICATE);
+/// ```
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub struct X509<'a> {
+    /// The certificate in bytes
+    bytes: &'a [u8],
+
+    /// The length of the certificate
+    len: usize,
+}
+
+impl<'a> X509<'a> {
+    /// Reads certificate in pem format from bytes
+    ///
+    /// # Error
+    /// This function returns [TlsError::X509MissingNullTerminator] if the certificate
+    /// doesn't end with a null-byte.
+    pub fn pem(bytes: &'a [u8]) -> Result<Self, TlsError> {
+        if let Some(mut len) = X509::get_null(bytes) {
+            // Append 1 to the length, to account for the null byte
+            len = len + 1;
+            // Get a slice of only the certificate bytes
+            let slice = unsafe { core::slice::from_raw_parts(bytes.as_ptr(), len) };
+            Ok(Self { bytes: slice, len })
+        } else {
+            Err(TlsError::X509MissingNullTerminator)
+        }
+    }
+
+    /// Reads certificate in der format from bytes
+    ///
+    /// *Note*: This function assumes that the size of the size is the exact
+    /// length of the certificate
+    pub fn der(bytes: &'a [u8]) -> Self {
+        Self {
+            bytes,
+            len: bytes.len(),
+        }
+    }
+
+    /// Returns the bytes of the certificate
+    pub fn data(&self) -> &'a [u8] {
+        self.bytes
+    }
+
+    /// Returns the length of the certificate
+    pub(crate) fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Returns a pointer to the data for parsing
+    pub(crate) fn as_ptr(&self) -> *const c_uchar {
+        self.data().as_ptr().cast()
+    }
+
+    /// Gets the first null byte in a slice
+    fn get_null(bytes: &[u8]) -> Option<usize> {
+        bytes.iter().position(|&byte| byte == 0)
```

Comment:
wouldn't it be good enough to check the last byte?

Reply:
Same as https://github.com/esp-rs/esp-mbedtls/pull/6#discussion_r1183398663